pkg/discovery/discovery_manager.go

```diff
@@ -292,25 +300,21 @@ func (m *Manager) AllGroups() map[string][]*target.Group {
 		m.metrics.discoveredTargets.WithLabelValues(setName).Set(float64(v))
 	}
 
-	m.updateProcessLabels(tSets)
-
-	return tSets
-}
-
-func (m *Manager) updateProcessLabels(tSets map[string][]*target.Group) {
+	// Update process labels.
+	pidLabels := map[int]model.LabelSet{}
 	for _, groups := range tSets {
 		for _, group := range groups {
-			for _, pid := range group.Pids {
+			for _, pid := range group.PIDs {
 				// Overwrite the information we have here with the latest.
-				// TODO: handle deletions.
-				m.pidLabels[pid] = group.Labels
+				pidLabels[pid] = group.Labels
 			}
 		}
 	}
-}
+	// Just swap the map with the new one. Considering how frequently we profile,
+	// we shouldn't receive anything when the discoverer reconcile labels.
+	m.pidLabels = pidLabels
```

Comment:
Could this potentially overwrite previous values we care about?

I am thinking of this:
1. The provider starts, and provides the currently running k8s pods
2. Then a new pod starts, my understanding is that we would get an update just for it
3. We will wipe the data stored in `m.pidLabels` for the pids stored in 1)

Reply:
I thought about this and decided not to spend too much time on it. This could happen for a single profile run where the process disappears, and we receive a profile after that. The downside of the previous approach we never delete anything from the map, and that meant we ever only aggregated, basically, it was a leak.

I'm happy to introduce some sort of time-based strategy to clean if the metadata has not been accessed for a time. What do you think?

Reply:
> I'm happy to introduce some sort of time-based strategy to clean if the metadata has not been accessed for a time. What do you think?

Sounds great, it's what I was thinking to do at some point to avoid the unbounded growth of the map as you mentioned 😄 . We can do this in another PR if you prefer

Reply:
I'll address it on this one :+1:

Reply:
@javierhonduco PTAL https://github.com/parca-dev/parca-agent/pull/663/commits/5bdec23d1612a64a09860f72fd7e029aa9e93a93